sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/implementation/PartitionBasedLoadBalancer.java

```diff
@@ -0,0 +1,352 @@
+            ownerPartitionMap.putIfAbsent(this.ownerId, new ArrayList<>());
+
+            /*
+             * Find the minimum number of partitions every event processor should own when the load is
+             * evenly distributed.
+             */
+            int numberOfActiveEventProcessors = ownerPartitionMap.size();
+            logger.info("Number of active event processors {}", ownerPartitionMap.size());
+
+            int minPartitionsPerEventProcessor = numberOfPartitions / numberOfActiveEventProcessors;
+
+            /*
+             * If the number of partitions in Event Hub is not evenly divisible by number of active event processors,
+             * a few Event Processors may own 1 additional partition than the minimum when the load is balanced. Calculate
+             * the number of event processors that can own additional partition.
+             */
+            int numberOfEventProcessorsWithAdditionalPartition = numberOfPartitions % numberOfActiveEventProcessors;
+
+            logger.info("Expected min partitions per event processor = {}, expected number of event "
+                + "processors with additional partition = {}", minPartitionsPerEventProcessor, numberOfEventProcessorsWithAdditionalPartition);
+
+            if (isLoadBalanced(minPartitionsPerEventProcessor, numberOfEventProcessorsWithAdditionalPartition,
+                ownerPartitionMap)) {
+                // If the partitions are evenly distributed among all active event processors, no change required.
+                logger.info("Load is balanced");
+                return;
+            }
+
+            if (!shouldOwnMorePartitions(minPartitionsPerEventProcessor, ownerPartitionMap)) {
+                // This event processor already has enough partitions and shouldn't own more.
+                logger.info("This event processor owns {} partitions and shouldn't own more",
+                    ownerPartitionMap.get(ownerId));
+                return;
+            }
+
+            // If we have reached this stage, this event processor has to claim/steal ownership of at least 1 more partition
+            logger.info(
+                "Load is unbalanced and this event processor should own more partitions");
+            /*
+             * If some partitions are unclaimed, this could be because an event processor is down and
+             * it's partitions are now available for others to own or because event processors are just
+             * starting up and gradually claiming partitions to own or new partitions were added to Event Hub.
+             * Find any partition that is not actively owned and claim it.
+             *
+             * OR
+             *
+             * Find a partition to steal from another event processor. Pick the event processor that has owns the highest
+             * number of partitions.
+             */
+            String partitionToClaim = partitionIds.parallelStream()
```

Comment:
Just curious, why are we using `parallelStream()` instead of regular `stream()`? are we making any service calls/ IO calls ?